mobile/files.go

```diff
@@ -0,0 +1,494 @@
+package mobile
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	ipld "gx/ipfs/QmZtNq8dArGfnpCZfx2pUNY7UcjGhVp5qqwQ4hH6mpTMRQ/go-ipld-format"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+
+	"github.com/golang/protobuf/ptypes"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/textileio/textile-go/pb"
+
+	"github.com/textileio/textile-go/core"
+	"github.com/textileio/textile-go/ipfs"
+	"github.com/textileio/textile-go/mill"
+	"github.com/textileio/textile-go/repo"
+	"github.com/textileio/textile-go/schema"
+)
+
+// FileData is a wrapper around a file data url
+type FileData struct {
+	Url string `json:"url"`
+}
+
+// AddSchema adds a new schema via schema mill
+func (m *Mobile) AddSchema(jsonstr string) (string, error) {
+	if !m.node.Started() {
+		return "", core.ErrStopped
+	}
+	added, err := m.addSchema(jsonstr)
+	if err != nil {
+		return "", err
+	}
+
+	return toJSON(added)
+}
+
+// PrepareFiles processes a file by path for a thread, but does NOT share it
+func (m *Mobile) PrepareFiles(path string, threadId string) (string, error) {
+	thrd := m.node.Thread(threadId)
+	if thrd == nil {
+		return "", core.ErrThreadNotFound
+	}
+
+	if thrd.Schema == nil {
+		return "", core.ErrThreadSchemaRequired
+	}
+
+	var result interface{}
+
+	mil, err := getMill(thrd.Schema.Mill, thrd.Schema.Opts)
+	if err != nil {
+		return "", err
+	}
+	if mil != nil {
+		conf, err := m.getFileConfig(mil, path, "")
+		if err != nil {
+			return "", err
+		}
+
+		added, err := m.node.AddFile(mil, *conf)
+		if err != nil {
+			return "", err
+		}
+		result = &added
+
+	} else if len(thrd.Schema.Links) > 0 {
+		dir := make(map[string]*repo.File)
+
+		// determine order
+		steps, err := schema.Steps(thrd.Schema.Links)
+		if err != nil {
+			return "", err
+		}
+
+		// send each link
+		for _, step := range steps {
+			mil, err := getMill(step.Link.Mill, step.Link.Opts)
+			if err != nil {
+				return "", err
+			}
+			var conf *core.AddFileConfig
+
+			if step.Link.Use == schema.FileTag {
+				conf, err = m.getFileConfig(mil, path, "")
+				if err != nil {
+					return "", err
+				}
+
+			} else {
+				if dir[step.Link.Use] == nil {
+					return "", errors.New(step.Link.Use + " not found")
+				}
+				conf, err = m.getFileConfig(mil, path, dir[step.Link.Use].Hash)
+				if err != nil {
+					return "", err
+				}
+			}
+			added, err := m.node.AddFile(mil, *conf)
+			if err != nil {
+				return "", err
+			}
+			dir[step.Name] = added
+		}
+		result = &dir
+
+	} else {
+		return "", schema.ErrEmptySchema
+	}
+
+	return toJSON(result)
+}
+
+// Callback is used for asyc methods (payload is a protobuf)
+type Callback interface {
+	Call([]byte, error)
+}
+
+// PrepareFilesAsync is the async flavor of PrepareFiles
+func (m *Mobile) PrepareFilesAsync(path string, threadId string, cb Callback) {
```

Comment:
@asutula I added an async version of `PrepareFiles` here which takes a callback interface. Let me know when you're around and we can talk through this and other changes.